hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -27,7 +27,59 @@ class Field[T: TypeInfo](classBuilder: ClassBuilder[_], val name: String) {
     Code(obj, v, Code(new FieldInsnNode(PUTFIELD, classBuilder.name, name, desc)))
 }
 
-class ClassBuilder[C](val name: String) {
+class ClassesBytes(classesBytes: Array[(String, Array[Byte])]) extends Serializable {
+  @transient @volatile var loaded: Boolean = false
+
+  def load(): Unit = {
+    if (!loaded) {
+      synchronized {
+        if (!loaded) {
+          classesBytes.foreach { case (n, bytes) =>
+            try {
+              HailClassLoader.loadOrDefineClass(n, bytes)
+            } catch {
+              case e: Exception =>
+                FunctionBuilder.bytesToBytecodeString(bytes, FunctionBuilder.stderrAndLoggerErrorOS)
+                throw e
+            }
+          }
+        }
+        loaded = true
+      }
+    }
+  }
+}
+
+class ModuleBuilder {
+  val classes = new mutable.ArrayBuffer[ClassBuilder[_]]()
+
+  def newClass[C](name: String = null): ClassBuilder[C] = {
+    val c = new ClassBuilder[C](this, name)
+    classes += c
+    c
+  }
+
+  var classesBytes: ClassesBytes = _
+
+  def classesBytes(print: Option[PrintWriter] = None): ClassesBytes = {
+    if (classesBytes == null) {
+      classesBytes = new ClassesBytes(
+        classes
+          .iterator
+          .map(c => (c.name.replace("/", "."), c.classAsBytes(print)))
+          .toArray)
+
+    }
+    classesBytes
+  }
+}
+
+class ClassBuilder[C](
+  module: ModuleBuilder,
+  val name: String = null) {
+  // FIXME use newClass
```

Comment:
Is this a `FIXME` before merging or one for sometime in the future?

Reply:
Let's say sometime in the future.

Reply:
I'm going to be making a bunch more passes over this code.